README.md
# `switchenv`: An environment manger for bash
In my analysis work, I will frequently put auth credentials into environment
variables.  This allows me to check my code into Github without divulging any
secrets.  So, for example, I might have code in a Jupyter notebook that looks
something like
```python
import os
from my_module import get_database_connection

connection = get_database_connection(
    port=os.environ('PGPORT'),
    password=os.environ('PGPASSWORD'),
    user=os.environ('PGUSER'),
    database=os.environ('PGDATABASE'),
    host=os.environ('PGHOST'),
)
```
The database I connect to will be completely determined by the environment
variables I have defined.

`switchenv` gives me a way to easily navigate between different environments so
that, for example, I can quickly switch between development and production
databases.

# Install
`switchenv` is written in python.  You can install it with
```bash
pip install switchenv
```

# Use Case
Imagine I have the following bash files I can source in order to set up my bash
environment the way I'd like.  Typically I would just run `source
rc_development_db.sh` before running my code in order to set up my dev
environment.  However, I have to be very careful about where I place my rc files
so that I don't accidentally check them into Github.  Furthermore, it can be
annoying to keep track of these files when I'd like to reuse them for different
projects.  This is where `switchenv` comes in.

`rc_development_db.sh`
```bash
export MY_MESSAGE="You are in dev profile"
export PGPORT=5432
export PGPASSWORD=my_dev_password
export PGUSER=my_dev_username
export PGDATABASE=my_dev_database
export PGHOST=my_dev_host
```

`rc_production_db.sh`
```bash
export MY_MESSAGE="You are in prod profile"
export PGPORT=5432
export PGPASSWORD=my_prod_password
export PGUSER=my_prod_username
export PGDATABASE=my_prod_database
export PGHOST=my_prod_host
```

`rc_bash_functions.sh`
```bash

print_message () {
   echo "Current message is ${MY_MESSAGE}"
}
```

# Setting up `switchenv`
## Basic Setup
Below is copy-paste from an interactive bash session showing how to set up the
`switchenv` workflow.


```bash
bash>
bash> # Make sure I'm in a directory with the rc files I want
bash> ls *.sh
rc_development_db.sh  rc_production_db.sh  rc_bash_functions.sh
bash>
bash> # Load my rc scripts into switchenv giving them profile names
bash> switchenv add -p dev -f ./rc_development_db.sh
bash> switchenv add -p prod -f ./rc_production_db.sh
bash>
bash> # Profiles can hold any legal bash code, including function definitions.
bash> switchenv add -p func -f ./rc_bash_functions.sh
bash>
bash> # Show list of stored profile names
bash> switchenv list
dev
prod
func
bash>
bash> # Show contents of single profile
bash> switchenv show -p prod


#========================================
# prod
#========================================
export PGPORT=5432
export PGPASSWORD=my_prod_password
export PGUSER=my_prod_username
export PGDATABASE=my_prod_database
export PGHOST=my_prod_host
```

Under the hood, switchenv placed a json file in a hidden directory off of my
home directory.
```bash
~/.switchenv/profiles.json
```
This json file serves as the centralized data-store for all of my profile
information.

## Advanced Setup (Composed profiles)
I can also create composed profiles.  These profiles will source other named
profiles in the order they are specified.  Any changes I make to one of the
sourced profiles will automatically carry over to the composite profile.
Composed profiles can be nested.  So, a composed profile can have another composed
profile as one of its sub-profiles.  Here is an example of setting up a composed profile
```bash
bash>
bash> # Create a composed profile based on two other profiles
bash> switchenv compose -c prod_with_func -p prod -p func
bash>
bash> # Show all profiles including new composite profile
bash> switchenv list
dev
prod
prod_with_func -> ['prod', 'func']
func
bash>
```
Now when you list profiles, you can easily identify composed profiles and see
the order in which they will execute their sub-profiles.

# Navigating Between Environments with `switchenv`
Using `switchenv` involves interacting with a simple console-based UI, so it is
best illustrated using a gif.  Shown here is my admittedly sub-par screen
recording of how to use switchenv.

For a more thorough description use
```
switchenv --help
```
Or use the `switchenv` alias (because I hate typing) 
```
bash> sw --help

Usage: sw [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  add       Create a profile from file
  compose   Compose a new profile from existing profiles
  delete    Delete a profile
  examples  Show usage examples
  list      List all profile names
  show      Show contents of a single profile
  snapshot  Snapshot current env into a profile
  source    Drop into subshell with named profile (useful in scripts)
```

Just for my future reference, I made this recording by using the native OSX
screen recording feature to make a `.mov` file.  I then used
[Gif Brewary 3](https://apps.apple.com/us/app/gif-brewery-3-by-gfycat/id1081413713?mt=12)
to convert it to a `.gif` file by manually setting the speed to 150%  and
frames-per-sec to 6.  I let the software figure it out from there.


![Demo Gif](https://github.com/robdmc/switchenv/blob/master/images/switchenv_demo.gif)

If you feel like digging around under the hood to see what `switchenv` actually sourced
when activating your environment, you can look at
```bash
~/.switchenv/switchenvrc.sh
```
Your subshell was essentially invoked with the command
```bash
bash --init-file ~/.switchenv/switchenvrc.sh
```
You can manually execute this command in a new terminal window if you would like
an exact clone of your environment in a new console.

___
Projects by [robdmc](https://www.linkedin.com/in/robdecarvalho).
* [Pandashells](https://github.com/robdmc/pandashells) Pandas at the bash command line
* [Consecution](https://github.com/robdmc/consecution) Pipeline abstraction for Python
* [Behold](https://github.com/robdmc/behold) Helping debug large Python projects
* [Crontabs](https://github.com/robdmc/crontabs) Simple scheduling library for Python scripts
* [Switchenv](https://github.com/robdmc/switchenv) Manager for bash environments


CHANGELOG.md
## Changelog

#### 0.2.0
##### New features
 - Can now create composed profiles
 - Added the sw alias for switchenv at the command-line.

##### Under the hood
 - Changed format of json data store.  It is now versioned.

##### Bug fixes
 - Fixed bug in how bash prompt was mutated


switchenv/doit.py
#! /usr/bin/env python


import os
import sys
import json
import click
import shutil
from typing import Optional, List
import fuzzypicker
import textwrap


class cached_property(object):
    """
    This is a direct copy-paste of Django's cached property from
    https://github.com/django/django/blob/2456ffa42c33d63b54579eae0f5b9cf2a8cd3714/django/utils/functional.py#L38-50
    """
    def __init__(self, func):
        self.func = func

    def __get__(self, instance, type=None):
        if instance is None:
            return self
        res = instance.__dict__[self.func.__name__] = self.func(instance)
        return res


class SwitchEnv:
    # Define the file locations for persisting environments
    BLOB_DIR = os.path.realpath(os.path.expanduser('~/.switchenv'))
    BLOB_FILE = os.path.join(BLOB_DIR, 'profiles.json')
    TEMP_FILE = os.path.join(BLOB_DIR, '__temp_profiles__.json')
    BASH_RC_FILE = os.path.realpath(os.path.expanduser('~/.bashrc'))
    TEMP_RC_FILE = os.path.join(BLOB_DIR, 'switchenvrc.sh')

    def __init__(self):
        # Ensure directory structure every time class is instantiate4d
        os.makedirs(self.BLOB_DIR, exist_ok=True)

    def make_temp_rc_file(self, profile, code):

        input_code_lines = code.split('\n')
        pre_code_lines = []
        post_code_lines = [f'PS1="•{profile}•$PS1"']

        code_lines = pre_code_lines + input_code_lines + post_code_lines
        code = '\n'.join(code_lines)

        bashrc = ''
        if os.path.isfile(self.BASH_RC_FILE):
            with open(self.BASH_RC_FILE) as bashrc_file:
                bashrc = bashrc_file.read()

        bashrc = textwrap.dedent(f'{bashrc}\n{code}')
        bashrc = '\n'.join([f' {line}' for line in bashrc.split('\n') if line])
        with open(self.TEMP_RC_FILE, 'w') as temp_rc_file:
            temp_rc_file.write(bashrc)

    @cached_property
    def blob(self):
        """
        Returns the currently saved blob.  Empty dict if nothing saved.
        """
        return self._load_file(self.BLOB_FILE)

    @cached_property
    def keys(self):
        """
        Returns list of profile names
        """
        return sorted(self.blob.keys())

    def get_key(self):
        return fuzzypicker.picker(self.keys)

    def _reset(self):
        """
        Bust the cache for all cached properties
        """
        # try:
        #     del self.keys
        # except AttributeError:
        #     pass

        # try:
        #     del self.keys
        # except AttributeError:
        #     pass

        for attr in ['keys', 'blob']:
            try:
                delattr(self, attr)
            except AttributeError:
                pass

    def _load_file(self, file_name):
        """
        Load a json file with provided name.
        Returns blank dict if file doesn't exist
        """

        if os.path.isfile(file_name):
            with open(file_name, 'r') as data_file:
                blob = json.load(data_file)
        else:
            blob = {}
        return blob

    def _save_file(self, blob, file_name):
        """
        Saves a json blob to specified file_name
        """
        with open(file_name, 'w') as out_file:
            json.dump(blob, out_file)

    def _confirm_file_contents(self, blob, file_name):
        """
        Compares the contents of a blob with those
        of a saved file
        """
        saved_blob = self._load_file(file_name)
        return blob == saved_blob

    def save(self, blob):
        """
        Atomically save a blob to the canonical file_name
        """
        # Bust the cached property caches
        self._reset()

        # Save the blob to a temp file
        self._save_file(blob, self.TEMP_FILE)

        # If temp file contents match blob, overwrite standard blob
        # file with temp file
        if self._confirm_file_contents(blob, self.TEMP_FILE):
            shutil.move(self.TEMP_FILE, self.BLOB_FILE)

    def update(self, update_blob):
        """
        Add or update blob contents
        """
        blob = self.blob
        blob.update(update_blob)
        self.save(blob)

    def delete(self, keys):
        """
        Remove profiles from the blob
        """
        blob = self.blob
        for key in keys:
            blob.pop(key, None)
        self.save(blob)

    def show(self, key_list=None, template=None):
        """
        Show a single profile.  template is a .format() 
        template that will have access to the profile
        name using the 'key' variable.  Defaults to
        f'
        """
        if not key_list:
            key = self.get_key()
            key_list = [key]

        for key in key_list:
            self._show_single(key)

    def _show_single(self, key):
        if key not in self.keys:
            return

        print('\n')
        print('=' * 40)
        print(key)
        print('=' * 40)
        print(self.blob[key])

    @property
    def env(self):
        env = dict(os.environ)

        # OSX does a weird thing by setting this variable
        # It ends up screwing up execvpe, so zap it out of
        # the environment
        env.pop('__PYVENV_LAUNCHER__', None)
        env.pop('_', None)
        return env


def run_switch_env(profile: Optional[str] = None):
    swenv = SwitchEnv()
    if profile is None:
        profile = swenv.get_key()

    code = swenv.blob[profile]
    swenv.make_temp_rc_file(profile, code)

    commands = ['bash', '--init-file', swenv.TEMP_RC_FILE]

    os.execvpe('bash', commands, swenv.env)


@click.group()
def cli():
    pass


@cli.command(help='List all profile names')
def list():
    swenv = SwitchEnv()
    if not swenv.keys:
        print('No saved profies')
    for key in swenv.keys:
        print(key)


@cli.command(help='Show contents of a single profile')
@click.option('-p', '--profiles', multiple=True)
def show(profiles):
    swenv = SwitchEnv()
    if not profiles:
        profiles = [swenv.get_key()]
    swenv = SwitchEnv()
    swenv.show(key_list=profiles)


@cli.command(help='Delete a profile')
@click.option('-p', '--profiles', multiple=True)
def delete(profiles):
    swenv = SwitchEnv()
    if not profiles:
        profiles = [swenv.get_key()]

    confirm = input(f"\nDelete {profiles}\ny/n: ").lower()
    if confirm[0] != 'y':
        print('Nothing done')
        return

    initial_keys = set(swenv.keys)
    swenv.delete(keys=profiles)
    final_keys = set(swenv.keys)
    deleted_keys = initial_keys - final_keys
    if deleted_keys:
        print(f'Deleted profiles: {sorted(deleted_keys)}')


@cli.command(help='Create a profile from file')
@click.option('-p', '--profile_name', required=True)
@click.option('-f', '--file_name', required=True)
def add(profile_name, file_name):
    if not os.path.isfile(file_name):
        print(f"\nThe file '{file_name}' does not exist\n")
        sys.exit(1)

    with open(file_name, 'r') as code_file:
        code = code_file.read()

    blob = {profile_name: code}

    swenv = SwitchEnv()
    swenv.update(blob)


@cli.command(help='Snapshot current env into a profile')
@click.option('-p', '--profile_name', required=True)
def snapshot(profile_name):
    swenv = SwitchEnv()

    code_lines = []
    for key, val in swenv.env.items():
        code_lines.append(f'export {key}={val}')

    code = '\n'.join(code_lines)
    swenv.update({profile_name: code})


def main():
    if len(sys.argv) > 1:
        cli()
    else:
        run_switch_env()


if __name__ == '__main__':
    main()


switchenv/switchenv.py
#! /usr/bin/env python


import os
import sys
import json
import click
import shutil
from typing import Optional
import fuzzypicker
import textwrap
import warnings


class cached_property(object):
    """
    This is a direct copy-paste of Django's cached property from
    https://github.com/django/django/blob/2456ffa42c33d63b54579eae0f5b9cf2a8cd3714/django/utils/functional.py#L38-50
    """
    def __init__(self, func):
        self.func = func

    def __get__(self, instance, type=None):
        if instance is None:
            return self
        res = instance.__dict__[self.func.__name__] = self.func(instance)
        return res


class SwitchEnv:
    # Define the file locations for persisting environments
    BLOB_DIR = os.path.realpath(os.path.expanduser('~/.switchenv'))
    BLOB_FILE = os.path.join(BLOB_DIR, 'profiles.json')
    TEMP_FILE = os.path.join(BLOB_DIR, '__temp_profiles__.json')
    BASH_RC_FILE = os.path.realpath(os.path.expanduser('~/.bashrc'))
    TEMP_RC_FILE = os.path.join(BLOB_DIR, 'switchenvrc.sh')

    BLOB_VERSION = '1.0'

    def __init__(self):
        # Ensure directory structure every time class is instantiate4d
        os.makedirs(self.BLOB_DIR, exist_ok=True)

    def make_temp_rc_file(self, profile, code):

        input_code_lines = code.split('\n')
        # Save off the PS1 variable before anything can change it
        pre_code_lines = []
        if '__PSSWE__' not in self.env:
            pre_code_lines.append(' export __PSSWE__="$PS1"')

        # Add the profile name to the front of PS1
        post_code_lines = [
            f' export __PSSWE__="•{profile}•$__PSSWE__"',
            f' PS1="$__PSSWE__"'
        ]

        # Create lines that will store the current env
        env_lines = []
        for key, val in self.env.items():
            env_lines.append(f'export {key}="{val}"')
        env_code = '\n'.join(env_lines)

        # Build code from the stored profile
        pre_code = '\n'.join(pre_code_lines)
        post_code = '\n'.join(post_code_lines)
        code_lines = input_code_lines
        code = '\n'.join(code_lines)

        # Load in the user's bashrc file
        bashrc = ''
        if os.path.isfile(self.BASH_RC_FILE):
            with open(self.BASH_RC_FILE) as bashrc_file:
                bashrc = bashrc_file.read()

        # The order here is important.  The users .bashrc can reset
        # the path variable overriding any currently set envs vars.
        # So:
        #    1) Run their .bashrc
        #    2) export all current environment variables
        #    3) source the custom profile code
        bashrc = textwrap.dedent(f'\n{bashrc}\n{pre_code}\n{env_code}\n{code}\n{post_code}')
        bashrc = '\n'.join([f' {line}' for line in bashrc.split('\n') if line])
        with open(self.TEMP_RC_FILE, 'w') as temp_rc_file:
            temp_rc_file.write(bashrc)

    @cached_property
    def blob(self):
        """
        Returns the currently saved blob.  Empty dict if nothing saved.
        """
        blob = self._load_file(self.BLOB_FILE)
        version = blob.get('version', 'unversioned')
        if version != 'unversioned':
            return blob

        else:
            profiles = {}
            for profile_name, code in blob.items():
                entry = {'code': code, 'code_type': 'raw'}
                profiles[profile_name] = entry

            new_blob = {'version': self.BLOB_VERSION, 'profiles': profiles}
            return new_blob

    @cached_property
    def keys(self):
        """
        Returns list of profile names
        """
        return sorted(self.blob.get('profiles', {}).keys())

    @cached_property
    def items(self):
        """
        Returns list of profile names
        """
        return sorted(self.blob.get('profiles', {}).items())

    def get_key(self):
        key = fuzzypicker.picker(self.keys)
        if key is None:
            sys.exit(0)
        return key

    def _reset(self):
        """
        Bust the cache for all cached properties
        """
        for attr in ['keys', 'blob']:
            try:
                delattr(self, attr)
            except AttributeError:
                pass

    def _load_file(self, file_name):
        """
        Load a json file with provided name.
        Returns blank dict if file doesn't exist
        """

        if os.path.isfile(file_name):
            with open(file_name, 'r') as data_file:
                blob = json.load(data_file)
        else:
            blob = {}
        return blob

    def _save_file(self, blob, file_name):
        """
        Saves a json blob to specified file_name
        """
        with open(file_name, 'w') as out_file:
            json.dump(blob, out_file, indent=2)

    def _confirm_file_contents(self, blob, file_name):
        """
        Compares the contents of a blob with those
        of a saved file
        """
        saved_blob = self._load_file(file_name)
        return blob == saved_blob

    def get_code(self, profile_name):
        code_list = self._get_code_list(profile_name)
        return '\n'.join(code_list)

    def _get_code_list(self, profile_name, code_list=None):
        if code_list is None:
            code_list = []

        profiles = self.blob['profiles']

        entry = profiles.get(profile_name)
        if entry is None:
            print(f"No profile named '{profile_name}'")
            sys.exit(1)

        if entry['code_type'] == 'raw':
            code_list.append(f'# ------- switchenv starting code for profile: {profile_name}\n')
            code_list.append(entry['code'])

        elif entry['code_type'] == 'composed':
            for sub_profile_name in entry['code']:
                code_list = self._get_code_list(sub_profile_name, code_list=code_list)

        else:
            raise ValueError('Only code types allowed are raw and composed')

        return code_list

    def save(self, blob):
        """
        Atomically save a blob to the canonical file_name
        """
        # Bust the cached property caches
        self._reset()

        # Make sure the blob has the proper version
        blob['version'] = self.BLOB_VERSION

        # Save the blob to a temp file
        self._save_file(blob, self.TEMP_FILE)

        # If temp file contents match blob, overwrite standard blob
        # file with temp file
        if self._confirm_file_contents(blob, self.TEMP_FILE):
            shutil.move(self.TEMP_FILE, self.BLOB_FILE)
        else:
            warnings.warn('Warning.  File contents could not be verified.  Something went wrong with saving.')

    def update_composed(self, composed_profile_name, source_profile_names):
        blob = self.blob
        profiles = blob['profiles']
        entry = profiles.get(composed_profile_name, {'code_type': 'composed'})

        # Can only update same kind of code_type
        if entry['code_type'] != 'composed':
            raise RuntimeError('Trying to update a profile with wrong code type')

        entry['code'] = list(source_profile_names)

        # Save the entry to profiles
        profiles[composed_profile_name] = entry

        # Save profiles to blob
        blob['profiles'] = profiles

        # Save the blob
        self.save(blob)

    def update_raw(self, profile_name, code):
        """
        Add or update blob contents
        """
        # Get a reference to the blob
        blob = self.blob

        # All profile information held in the 'profiles' blob key
        profiles = blob.get('profiles', {})

        # Get the current entry or dict initialized with default code type
        entry = profiles.get(profile_name, {'code_type': 'raw'})

        # Can only update same kind of code_type
        if entry['code_type'] != 'raw':
            raise RuntimeError('Trying to update a profile with wrong code type')

        # Update the entry's code
        entry['code'] = code

        # Save the entry to profiles
        profiles[profile_name] = entry

        # Save profiles to blob
        blob['profiles'] = profiles

        # Save the blob
        self.save(blob)

    def delete(self, keys):
        """
        Remove profiles from the blob
        """
        blob = self.blob
        for key in keys:
            blob['profiles'].pop(key, None)
        self.save(blob)

    def show(self, key_list=None, template=None):
        """
        Show a single profile.  template is a .format()
        template that will have access to the profile
        name using the 'key' variable.  Defaults to
        f'
        """
        if not key_list:
            key = self.get_key()
            key_list = [key]

        for key in key_list:
            self._show_single(key)

    def _show_single(self, key):
        if key not in self.keys:
            return

        print('\n')
        print(f"#{'=' * 40}")
        print(f'# {key}')
        print(f"#{'=' * 40}")
        print(self.get_code(key))

    @property
    def env(self):
        env = dict(os.environ)

        # OSX does a weird thing by setting this variable
        # It ends up screwing up execvpe, so zap it out of
        # the environment
        # See: https://stackoverflow.com/questions/26323852/whats-the-meaning-of-pyvenv-launcher-environment-variable
        env.pop('__PYVENV_LAUNCHER__', None)
        env.pop('_', None)

        return env


def ensure_profiles_exist(swenv):
    if len(swenv.keys) == 0:
        print('\nNo saved profiles\n')
        sys.exit(1)


def run_switch_env(profile: Optional[str] = None):
    swenv = SwitchEnv()
    ensure_profiles_exist(swenv)

    if profile is None:
        profile = swenv.get_key()

    code = swenv.get_code(profile)
    swenv.make_temp_rc_file(profile, code)

    commands = ['bash', '--init-file', swenv.TEMP_RC_FILE]

    os.execvpe('bash', commands, swenv.env)


@click.group()
def cli():
    pass

@cli.command(help='Show usage examples')
def examples():
    text = textwrap.dedent("""

    # Snapshot the current environment as a profile
    switchenv snapshot -p my_snapshot_profile_name

    # Add an existing shell script as a profile
    switchenv add -p my_profile_name -f path/to/my_scrpt.sh

    # Create a composite profile
    switchenv compose -c my_composite_profile_name -p my_snapshot_profile_name -p my_profile_name

    # Show all profile names
    switchenv list

    # Show contents of a specific profile
    switchenv show                                # allows for fuzzysearch of profile name
    switchenv show -p profile1 [-p profile2... ]  # show a specific profile(s)

    # Drop into a subshell with a specific profile
    switchenv  # Will present you with a fuzzy searchable list of profiles

    # Drop into a named profile (useful for invoking in scripts)
    switchenv source -p profile_name

    # Delete profiles
    switchenv delete -p profile_name_1 [-p profile_name_2, ...]


    """)
    print(text)


@cli.command(name='list', help='List all profile names')
def list_profiles():
    swenv = SwitchEnv()
    ensure_profiles_exist(swenv)
    for key, profile in swenv.items:
        code_type = profile['code_type']
        if code_type == 'raw':
            print(key)
        elif code_type == 'composed':
            print(f'{key} -> {sorted(profile["code"])}')
        else:
            raise ValueError('unkown code_type')


@cli.command(help='Show contents of a single profile')
@click.option('-p', '--profiles', multiple=True)
def show(profiles):
    swenv = SwitchEnv()
    ensure_profiles_exist(swenv)
    if not profiles:
        profiles = [swenv.get_key()]
    swenv = SwitchEnv()
    swenv.show(key_list=profiles)

@cli.command(help='Drop into subshell with named profile (useful in scripts)')
@click.option('-p', '--profile', required=True)
def source(profile):
    run_switch_env(profile)



@cli.command(help='Delete a profile')
@click.option('-p', '--profiles', multiple=True)
def delete(profiles):
    swenv = SwitchEnv()
    ensure_profiles_exist(swenv)
    if not profiles:
        profiles = [swenv.get_key()]

    confirm = input(f"\nDelete {profiles}\ny/n: ").lower()
    if confirm[0] != 'y':
        print('Nothing done')
        return

    initial_keys = set(swenv.keys)
    swenv.delete(keys=profiles)
    final_keys = set(swenv.keys)
    deleted_keys = initial_keys - final_keys
    if deleted_keys:
        print(f'Deleted profiles: {sorted(deleted_keys)}')


@cli.command(help='Create a profile from file')
@click.option('-p', '--profile_name', required=True)
@click.option('-f', '--file_name', required=True)
def add(profile_name, file_name):
    if not os.path.isfile(file_name):
        print(f"\nThe file '{file_name}' does not exist\n")
        sys.exit(1)

    with open(file_name, 'r') as code_file:
        code = code_file.read()

    swenv = SwitchEnv()
    swenv.update_raw(profile_name, code)


@cli.command(help='Compose a new profile from existing profiles')
@click.option('-c', '--composed_profile_name', required=True, help='The name of the posed profile')
@click.option('-p', '--profiles', multiple=True, help='The name of the source profile')
def compose(composed_profile_name, profiles):
    if not profiles:
        print('You must supply at least one source profile')

    swenv = SwitchEnv()
    swenv.update_composed(composed_profile_name, profiles)


@cli.command(help='Snapshot current env into a profile')
@click.option('-p', '--profile_name', required=True)
def snapshot(profile_name):
    swenv = SwitchEnv()

    code_lines = []
    for key, val in swenv.env.items():
        code_lines.append(f'export {key}="{val}"')

    code = '\n'.join(code_lines)
    swenv.update_raw(profile_name, code)


def main():
    if len(sys.argv) > 1:
        cli()
    else:
        run_switch_env()


if __name__ == '__main__':
    main()
